
Exemple from 

Introduction to Stochastic Programming

John R. Birge • Franç̧̧ois Louveaux


# Formulation without uncertainties

a European farmer specializes in raising wheat, corn, and sugar beets 
on his 500 acres of land. During the winter, he wants to decide how much land to devote to each crop.


- **Wheat/Corn**

    - The farmer knows that at least 200 tons (T) of wheat and 240 T of corn are needed for cattle feed. 
    - These amounts can be raised on the farm or bought from a wholesaler. Any production in excess of the feeding requirement would be sold. 
    - Over the last decade, mean selling prices have been 170 and 150 (dollars/T) for wheat and corn. 
    - The purchase prices are 40% more than this due to the wholesalers margin and transportation costs.
    
- **Sugar beet**
    - The farmer expects to sell at 36 (dollars/T), 
    - the European Commission imposes a quota on sugar beet production. 
    - Any amount in excess of the quota can be sold only at 10 per ton. The farmer’s quota for next year is 6000 T.
    - the farmer knows that the mean yield on his land is roughly 20 T/acre,

- **The farmer** estimates that on his land
    - the mean yield is roughly 2.5, 3 and 20 T/acre (wheat,corn,sugar beet). 
    - the mean planting cost is roughly 150, 230, and 260 dollars/acre (wheat,corn,sugar beet). 


**Table 1: Data for farmer's problem.**

| Culture | Wheat| Corn | Sugar Beets |
| --- | --- | --- ||
| Yield (T) | 2.5| 3 | 20 |
| Planting cost  | 150| 230 | 60 |
| Sell price  | 170| 150 | 36 under 6000 T |
|  | |  | 10 above 6000 T |
| Purchase price  | 238| 210 | - |
| Minimum requirement (T) | 200| 240 | - |
---
Total available land: 500 acres


## Decision variables

- x1= acres of land devoted to wheat,
- x2= acres of land devoted to corn,
- x3= acres of land devoted to sugar beets,
- w1= tons of wheat sold,
- y1= tons of wheat purchased,
- w2= tons of corn sold,
- y2= tons of corn purchased,
- w3= tons of sugar beets sold at the favorable price,
- w4 = tons of sugar beets sold at the lower price.

## Constraints

- formulate the problem description as constaints on decision variables

## Objective

- formulate costs minimisation using problem datas and decision variables

---
---

**Solving the problem using OR-tools**

In [1]:
!pip install ortools

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.3/16.3 MB 33.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 302.4/302.4 KB 7.1 MB/s eta 0:00:00
  Attempting uninstall: protobuf
    Found existing installation: protobuf 3.19.6
    Uninstalling protobuf-3.19.6:
      Successfully uninstalled protobuf-3.19.6
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tensorflow 2.11.0 requires protobuf<3.20,>=3.9.2, but you have protobuf 4.22.0 which is incompatible.
tensorflow-metadata 1.12.0 requires protobuf<4,>=3.13, but you have protobuf 4.22.0 which is incompatible.
tensorboard 2.11.2 requires protobuf<4,>=3.9.2, but you have protobuf 4.22.0 which is incompatible.


In [2]:
from ortools.linear_solver import pywraplp


In [3]:
# Create a solver using the GLOP backend
solver = pywraplp.Solver('Minimize costs', pywraplp.Solver.GLOP_LINEAR_PROGRAMMING)

In [5]:
# Variables

x1 = solver.IntVar(0, solver.infinity(), 'acres of land devoted to wheat')
x2 = solver.IntVar(0, solver.infinity(), 'acres of land devoted to corn')
x3 = solver.IntVar(0, solver.infinity(), 'acres of land devoted to sugar beets')


w1 = solver.IntVar(0, solver.infinity(), 'tons of wheat sold')
y1 = solver.IntVar(0, solver.infinity(), 'tons of wheat purchased')


w2 = solver.IntVar(0, solver.infinity(), 'tons of corn sold')
y2 = solver.IntVar(0, solver.infinity(), 'tons of corn purchased')

w3 = solver.IntVar(0, solver.infinity(), 'tons of sugar beets sold at the favorable price')
w4 = solver.IntVar(0, solver.infinity(), 'tons of sugar beets sold at the lower price')


In [6]:
# Constraints

solver.Add(x1 + x2 + x3 <= 500)
solver.Add(2.5*x1 + y1 - w1 >= 200) 
solver.Add(3*x2 + y2 - w2 >= 240) 
solver.Add(w3 + w4 <= 20*x3) 
solver.Add(w3 <= 6000 )                  # acres of land available
# add the other constraints

<ortools.linear_solver.pywraplp.Constraint; proxy of <Swig Object of type 'operations_research::MPConstraint *' at 0x7fa524d28d80> >

In [7]:
# Objective

solver.Minimize(150*x1 + 230*x2 + 260*x3 + 238*y1 + 210*y2 - 170*w1 - 150*w2 - 36*w3 -10*w4)


In [9]:
status = solver.Solve()

# If an optimal solution has been found, print results
if status == pywraplp.Solver.OPTIMAL:
  print('================= Solution =================')
  print(f'Solved in {solver.wall_time():.2f} milliseconds in {solver.iterations()} iterations')
  print()
  print(f'Optimal power = {solver.Objective().Value()} total costs')
  print('land distribution:')
  print(f' - x1 = {x1.solution_value()}')
  print(f' - x2 = {x2.solution_value()}')
  print(f' - x3 = {x3.solution_value()}')
  print(f' - y1 = {y1.solution_value()}')
  print(f' - y2 = {y2.solution_value()}')
  print(f' - w1 = {w1.solution_value()}')
  print(f' - w2 = {w2.solution_value()}')
  print(f' - w3 = {w3.solution_value()}')
  print(f' - w4 = {w4.solution_value()}')
else:
  print('The solver could not find an optimal solution.')

================= Solution =================
Solved in 1523764.00 milliseconds in 0 iterations

Optimal power = -118599.99999999999 total costs
land distribution:
 - x1 = 119.99999999999996
 - x2 = 80.0
 - x3 = 300.0
 - y1 = 0.0
 - y2 = 0.0
 - w1 = 99.99999999999989
 - w2 = 0.0
 - w3 = 6000.0
 - w4 = 0.0


Fill in your result

**Table 2: Optimal solution based on expected yields.**

| Culture | Wheat| Corn | Sugar Beets |
| --- | --- | --- ||
| Surface (acres) | 120| 80 | 300 |
| Yield (T) | 300| 240 | 6000 |
| Sales (Dollars) | 17000| 0 | 216 000 |
| Purchase (Dollars) | 0| 0 | 0 |
---
Overall profit is 118 600 Dollars



# Formulation with uncertainties

## 3 scénarios 

The farmer assumes some correlation among the yields of the different crops. A very simplified representation of this would be 
- to assume that years are **good, fair, or bad** for all crops, 
- resulting in **above average, average, or below average** yields for all crops
- **above** and **below** average indicate a yield 20% above or below the mean yield given. 

The farmer wishes to know whether the optimal solution is sensitive to variations in yields. 

Above

In [14]:
# Constraints
# new_yield= old_yield*1.2  raise yield of 20%
solver.Add(x1 + x2 + x3 <= 500)
solver.Add(3*x1 + y1 - w1 >= 200) 
solver.Add(3.6*x2 + y2 - w2 >= 240) 
solver.Add(w3 + w4 <= 24*x3) 
solver.Add(w3 <= 6000 )                  # acres of land available
# add the other constraints

<ortools.linear_solver.pywraplp.Constraint; proxy of <Swig Object of type 'operations_research::MPConstraint *' at 0x7fa546a18120> >

In [15]:
status = solver.Solve()

# If an optimal solution has been found, print results
if status == pywraplp.Solver.OPTIMAL:
  print('================= Solution =================')
  print(f'Solved in {solver.wall_time():.2f} milliseconds in {solver.iterations()} iterations')
  print()
  print(f'Optimal power = {solver.Objective().Value()} total costs')
  print('land distribution:')
  print(f' - x1 = {x1.solution_value()}')
  print(f' - x2 = {x2.solution_value()}')
  print(f' - x3 = {x3.solution_value()}')
  print(f' - y1 = {y1.solution_value()}')
  print(f' - y2 = {y2.solution_value()}')
  print(f' - w1 = {w1.solution_value()}')
  print(f' - w2 = {w2.solution_value()}')
  print(f' - w3 = {w3.solution_value()}')
  print(f' - w4 = {w4.solution_value()}')
else:
  print('The solver could not find an optimal solution.')

================= Solution =================
Solved in 6288937.00 milliseconds in 0 iterations

Optimal power = -59950.0 total costs
land distribution:
 - x1 = 100.00000000000001
 - x2 = 25.00000000000004
 - x3 = 374.99999999999994
 - y1 = 0.0
 - y2 = 179.99999999999991
 - w1 = 0.0
 - w2 = 0.0
 - w3 = 6000.0
 - w4 = 0.0


# Below


In [10]:
# Constraints
# new_yield= old_yield*0.8  decrease yield of 20%
solver.Add(x1 + x2 + x3 <= 500)
solver.Add(2*x1 + y1 - w1 >= 200) 
solver.Add(2.4*x2 + y2 - w2 >= 240) 
solver.Add(w3 + w4 <= 16*x3) 
solver.Add(w3 <= 6000 )                  # acres of land available
# add the other constraints

<ortools.linear_solver.pywraplp.Constraint; proxy of <Swig Object of type 'operations_research::MPConstraint *' at 0x7fa524d283c0> >

In [11]:
status = solver.Solve()

# If an optimal solution has been found, print results
if status == pywraplp.Solver.OPTIMAL:
  print('================= Solution =================')
  print(f'Solved in {solver.wall_time():.2f} milliseconds in {solver.iterations()} iterations')
  print()
  print(f'Optimal power = {solver.Objective().Value()} total costs')
  print('land distribution:')
  print(f' - x1 = {x1.solution_value()}')
  print(f' - x2 = {x2.solution_value()}')
  print(f' - x3 = {x3.solution_value()}')
  print(f' - y1 = {y1.solution_value()}')
  print(f' - y2 = {y2.solution_value()}')
  print(f' - w1 = {w1.solution_value()}')
  print(f' - w2 = {w2.solution_value()}')
  print(f' - w3 = {w3.solution_value()}')
  print(f' - w4 = {w4.solution_value()}')
else:
  print('The solver could not find an optimal solution.')

================= Solution =================
Solved in 5746190.00 milliseconds in 2 iterations

Optimal power = -59950.0 total costs
land distribution:
 - x1 = 100.0
 - x2 = 25.0
 - x3 = 375.0
 - y1 = 0.0
 - y2 = 179.99999999999997
 - w1 = 0.0
 - w2 = 0.0
 - w3 = 6000.0
 - w4 = 0.0


- Reproduce Table 1 for the above average and below average

*answer*

- Run the two optimizations present results in tables for 
  - above average and 
  - below average yields

*answer*


- Are the optimal solutions sensitive to changes in yields
- What is the main issue comparing the three scenarios

In [18]:
# Constraints
# new_yield= old_yield*0.8  decrease yield of 20%
solver.Add(x1 + x2 + x3 <= 500)
solver.Add(2*x1 + y1 - w1 >= 200) 
solver.Add(2.5*x1 + y1 - w1 >= 200)
solver.Add(3*x1 + y1 - w1 >= 200)
solver.Add(2.4*x2 + y2 - w2 >= 240) 
solver.Add(3*x2 + y2 - w2 >= 240) 
solver.Add(3.6*x2 + y2 - w2 >= 240) 
solver.Add(w3 + w4 <= 16*x3)
solver.Add(w3 + w4 <= 20*x3) 
solver.Add(w3 + w4 <= 24*x3)  
solver.Add(w3 <= 6000 )                  # acres of land available
# add the other constraints

<ortools.linear_solver.pywraplp.Constraint; proxy of <Swig Object of type 'operations_research::MPConstraint *' at 0x7fa524c52b70> >

In [19]:
status = solver.Solve()

# If an optimal solution has been found, print results
if status == pywraplp.Solver.OPTIMAL:
  print('================= Solution =================')
  print(f'Solved in {solver.wall_time():.2f} milliseconds in {solver.iterations()} iterations')
  print()
  print(f'Optimal power = {solver.Objective().Value()} total costs')
  print('land distribution:')
  print(f' - x1 = {x1.solution_value()}')
  print(f' - x2 = {x2.solution_value()}')
  print(f' - x3 = {x3.solution_value()}')
  print(f' - y1 = {y1.solution_value()}')
  print(f' - y2 = {y2.solution_value()}')
  print(f' - w1 = {w1.solution_value()}')
  print(f' - w2 = {w2.solution_value()}')
  print(f' - w3 = {w3.solution_value()}')
  print(f' - w4 = {w4.solution_value()}')
else:
  print('The solver could not find an optimal solution.')

================= Solution =================
Solved in 8073590.00 milliseconds in 3 iterations

Optimal power = -59950.00000000003 total costs
land distribution:
 - x1 = 100.00000000000001
 - x2 = 25.000000000000053
 - x3 = 374.99999999999994
 - y1 = 0.0
 - y2 = 179.99999999999983
 - w1 = 0.0
 - w2 = 0.0
 - w3 = 6000.0
 - w4 = 0.0


answer